File: api/extract.js
// api/extract.js
// Receives page texts from browser, calls OpenAI, returns product JSON.
// JS instead of Python: cold start ~200ms vs ~3s, fits Vercel free 10s limit.

export const config = { api: { bodyParser: { sizeLimit: '4mb' } } }

// Column schema per category
const COLUMNS = {
  default:    ['page_no','product_code','product_name','size','material','finish','color_family','texture_type','collection','brand','serial_no','notes'],
  tiles:      ['page_no','product_code','product_name','size','finish','material','color_family','texture_type','collection','brand','thickness','notes'],
  laminates:  ['page_no','product_code','product_name','size','thickness','finish','material','color_family','texture_type','collection','brand','notes'],
  panels:     ['page_no','product_code','product_name','size','finish','material','color_family','texture_type','thickness','collection','brand','application','notes'],
  louvers:    ['page_no','product_code','product_name','size','profile','finish','material','color_family','collection','brand','application','notes'],
  wallpapers: ['page_no','product_code','product_name','size','finish','material','color_family','pattern_type','collection','brand','notes'],
  quartz:     ['page_no','product_code','product_name','size','finish','color_family','thickness','collection','brand','edge_profile','notes'],
}

function buildPrompt(category, columns, pages) {
  const colStr   = columns.join(', ')
  const pageText = pages.map(p => `\n\n--- PAGE ${p.page_no} ---\n${(p.text || '').slice(0, 3000)}`).join('')

  return `You are extracting product data from a building materials catalogue (${category}).
Analyse all pages below and return ONLY a JSON object with a "products" array.

Each product must have EXACTLY these fields (use "" for unknown values):
${colStr}

RULES:
1. Only include pages with actual product listings that have a product code.
2. SKIP: cover pages, index pages, room/lifestyle photos, brand story pages, any page without a product code.
3. product_code = exact code printed (e.g. LPR-514, TL-2201). Skip page if no code found.
4. size = full dimension string as printed (e.g. "2400 x 1200 MM").
5. If one page has multiple products, return one row per product.
6. Keep values concise — one line each.
7. Return ONLY valid JSON, no markdown, no explanation.

PAGE TEXTS:${pageText}`
}

export default async function handler(req, res) {
  if (req.method !== 'POST') return res.status(405).end()

  const { category = 'default', pages = [] } = req.body || {}
  if (!pages.length) return res.status(400).json({ error: 'No pages provided' })

  const apiKey = process.env.OPENAI_API_KEY
  const model  = process.env.OPENAI_MODEL || 'gpt-4o-mini'
  if (!apiKey) return res.status(500).json({ error: 'OPENAI_API_KEY not set' })

  const columns = COLUMNS[category.toLowerCase()] || COLUMNS.default

  // For large PDFs, only send first 25 pages to stay within timeout
  // User can re-run with page range if needed
  const pagesToSend = pages.slice(0, 25)

  try {
    const prompt = buildPrompt(category, columns, pagesToSend)

    const aiRes = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type':  'application/json',
        'Authorization': `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model,
        temperature:     0,
        response_format: { type: 'json_object' },
        messages: [
          { role: 'system', content: 'Return only valid JSON with a "products" array. No markdown.' },
          { role: 'user',   content: prompt },
        ],
      }),
    })

    const aiText = await aiRes.text()

    if (!aiRes.ok) {
      let errMsg = `OpenAI error ${aiRes.status}`
      try { errMsg = JSON.parse(aiText).error?.message || errMsg } catch {}
      return res.status(500).json({ error: errMsg })
    }

    let aiData
    try { aiData = JSON.parse(aiText) } catch {
      return res.status(500).json({ error: `OpenAI returned non-JSON: ${aiText.slice(0, 200)}` })
    }

    const content = aiData.choices?.[0]?.message?.content || ''
    let products = []
    try {
      const parsed = JSON.parse(content)
      products = parsed.products || []
    } catch {
      return res.status(500).json({ error: `Could not parse products from AI response: ${content.slice(0, 300)}` })
    }

    return res.status(200).json({
      products,
      count:      products.length,
      pages_sent: pagesToSend.length,
      total_pages: pages.length,
    })

  } catch (err) {
    return res.status(500).json({ error: String(err.message || err) })
  }
}

File: api/extract.py
# api/extract.py
# ══════════════════════════════════════════════════
# Receives page texts from the browser, calls OpenAI
# once, returns a JSON array of product rows.
# Text-only — no images, no Drive, no complexity.
# Fits comfortably within Vercel's 60s limit.
# ══════════════════════════════════════════════════

import os, json, traceback
from http.server import BaseHTTPRequestHandler
from openai import OpenAI

# ── Column schema per category ──────────────────────────────
# These become the CSV headers OpenAI must populate.

COLUMNS = {
    "default": [
        "page_no", "product_code", "product_name", "size",
        "material", "finish", "color_family", "texture_type",
        "collection", "brand", "serial_no", "notes",
    ],
    "tiles": [
        "page_no", "product_code", "product_name", "size",
        "finish", "material", "color_family", "texture_type",
        "collection", "brand", "thickness", "surface_area", "notes",
    ],
    "laminates": [
        "page_no", "product_code", "product_name", "size",
        "thickness", "finish", "material", "color_family",
        "texture_type", "collection", "brand", "notes",
    ],
    "panels": [
        "page_no", "product_code", "product_name", "size",
        "finish", "material", "color_family", "texture_type",
        "thickness", "collection", "brand", "application", "notes",
    ],
    "louvers": [
        "page_no", "product_code", "product_name", "size",
        "profile", "finish", "material", "color_family",
        "collection", "brand", "application", "notes",
    ],
    "wallpapers": [
        "page_no", "product_code", "product_name", "size",
        "finish", "material", "color_family", "pattern_type",
        "collection", "brand", "notes",
    ],
    "quartz": [
        "page_no", "product_code", "product_name", "size",
        "finish", "color_family", "thickness",
        "collection", "brand", "edge_profile", "notes",
    ],
}


def build_prompt(category, columns, pages):
    cols_str  = ", ".join(columns)
    pages_str = ""
    for p in pages:
        pages_str += f"\n\n--- PAGE {p['page_no']} ---\n{p['text'][:3000]}"

    return f"""You are extracting product data from a building materials catalogue ({category}).
Analyse all pages below and return ONLY a JSON object with a "products" array.

Each product must have EXACTLY these fields (use "" for unknown values):
{cols_str}

RULES:
1. Only include pages that contain actual product listings with a product code.
2. Skip: cover pages, index pages, application/room photos, brand story pages, spec-only pages with no product code.
3. product_code must be the exact code printed (e.g. LPR-514, TL-2201, WP-38). Skip if no code found.
4. size = full dimension string as printed (e.g. "2400 x 1200 MM", "600 x 600 MM").
5. If one page has multiple products, return one row per product.
6. Keep all values concise — one line per field.
7. Return ONLY valid JSON. No markdown. No explanation.

PAGE TEXTS:
{pages_str}
"""


def extract(ai, category, columns, pages, model):
    prompt = build_prompt(category, columns, pages)
    resp = ai.chat.completions.create(
        model=model,
        temperature=0,
        response_format={"type": "json_object"},
        messages=[
            {"role": "system", "content": "Return only valid JSON with a 'products' array. No markdown."},
            {"role": "user",   "content": prompt},
        ],
        timeout=45,
    )
    raw = resp.choices[0].message.content
    data = json.loads(raw)
    return data.get("products", [])


class handler(BaseHTTPRequestHandler):
    def log_message(self, *a): pass

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body   = json.loads(self.rfile.read(length))

        category   = (body.get("category") or "default").lower().strip()
        pages      = body.get("pages") or []
        model      = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")
        ai         = OpenAI(api_key=os.environ["OPENAI_API_KEY"])

        columns    = COLUMNS.get(category, COLUMNS["default"])

        # If PDF is very long, split into chunks of 30 pages and merge
        CHUNK = 30
        all_products = []

        try:
            if len(pages) <= CHUNK:
                all_products = extract(ai, category, columns, pages, model)
            else:
                for i in range(0, len(pages), CHUNK):
                    chunk = pages[i : i + CHUNK]
                    chunk_result = extract(ai, category, columns, chunk, model)
                    all_products.extend(chunk_result)

            return self._send(200, {"products": all_products, "count": len(all_products)})

        except Exception as e:
            return self._send(500, {"error": str(e), "trace": traceback.format_exc()})

    def _send(self, code, body):
        payload = json.dumps(body).encode()
        self.send_response(code)
        self.send_header("Content-Type",  "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(payload)
